indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java

```diff
@@ -120,23 +135,46 @@ public DimFilter getDimensionsFilter()
     return metrics;
   }
 
-  public void setTaskToolbox(TaskToolbox taskToolbox)
-  {
-    this.taskToolbox = taskToolbox;
-  }
-
   @Override
   public Firehose connect(InputRowParser inputRowParser, File temporaryDirectory) throws ParseException
   {
     log.info("Connecting firehose: dataSource[%s], interval[%s]", dataSource, interval);
 
-    Preconditions.checkNotNull(taskToolbox, "taskToolbox is not set");
-
     try {
-      final List<DataSegment> usedSegments = taskToolbox
-          .getTaskActionClient()
-          .submit(new SegmentListUsedAction(dataSource, interval, null));
-      final Map<DataSegment, File> segmentFileMap = taskToolbox.fetchSegments(usedSegments);
+      // This call used to use the TaskActionClient, so for compatibility we use the same retry configuration
+      // as TaskActionClient.
+      final RetryPolicy retryPolicy = retryPolicyFactory.makeRetryPolicy();
+      List<DataSegment> usedSegments;
+      while (true) {
```

Comment:
This looks good to me, but I feel like we can use `RetryUtils.retry()` to remove duplicate codes even though the retry logic is slightly different. It's up to you.

Reply:
It seems like the logic in `RetryUtils` is somewhat different and not configurable. If we want to keep the compatibility I'd rather keep it this way. A nice little project might be to move RetryPolicy and friends into `core` along with `RetryUtils` and then have a version of `RetryUtils.retry` that takes a `RetryPolicyConfig` or something — actually maybe just dropping `RetryPolicy` and `RetryPolicyFactory` entirely? But I feel like this PR refactors enough already.

Reply:
I agree. Integrating `RetryPolicy` and `RetryUtils` sounds good, but should be done in a separate PR.